SqueakJS V2 (WIP)

codefrau opened the pull request:
Goals of V2: see [README](https://github.com/codefrau/SqueakJS/blob/v2/README.md)

This PR mainly exists as a place for discussion.

codefrau committed:
Readme for V2

codefrau committed:
More README

codefrau committed:
Update README.md

LinqLover commented:
Interesting, thanks for sharing! Am I qualified to ask one or two questions? :-)

1. About: 

   > Still need to decide between named inst vars (using inst var names from image) or suffixed (like p0, p1, ...)

   Just curious, are instance variable names a concept that the VM should be aware of at all? Or could I in theory create a new subclass manual in my image, do not assign names to the variables but only address them via indexes, and complain if SqueakJS cannot handle this class? Also, could there be valid Smalltalk variable names that are not allowed in JavaScript (I know the ClassBuilder is pretty restrictive but maybe it would not have to be)?

2. What order of magnitude of speed-ups do you think each of these design changes will yield? Is there any bottleneck analysis publicly available? 

Sorry if this is out of topic ...

codefrau committed:
Add link to JIT experiments

codefrau commented:
> are instance variable names a concept that the VM should be aware of at all? Or could I in theory create a new subclass manual in my image, do not assign names to the variables but only address them via indexes, and complain if SqueakJS cannot handle this class? 

If no suitable inst var name string can be found, we can easily generate a name like `p0`, `p1` etc. Same when sanitizing names.

> 2. What order of magnitude of speed-ups do you think each of these design changes will yield? Is there any bottleneck analysis publicly available?

See [here](https://squeak.js.org/docs/jit.md.html), in particular [this](https://squeak.js.org/docs/jit.md.html#sketchingajit/exception-basedschemewithblocksandnon-localreturns/performanceestimate) section. I haven't done a complete analysis though. This is still a hobby for me ;)

I believe we can overall get a 10x improvement in send speed. That would get us to within a magnitude of Cog.

bpieber commented:
I love the fact that SqueakJS runs a recent Cuis image. It pauses often when using it, though. I wonder if porting the VectorEnginePlugin would help and be feasible? Juan also mentioned the possibility to implement the VectorCanvas using Web APIs. Maybe that would be easier?

codefrau commented:
@bpieber This is independent of v2. I just opened #169 for it, please continue this discussion there.

codefrau committed:
Update README.md

codefrau committed:
we can use modern JS now

LinqLover commented:
@codefrau Very interesting read, thank you! :-)

codefrau committed:
Rename fromImage to fromBits

codefrau committed:
use actual inst var names

This uses Smalltalk inst var names for property names prefixed with a `$` (e.g. a Point would use `{ $x: 3, $y: 4 }` in JS). Indexable fields are in an array named `$$` (e.g. in Context you would have `{$sender: ..., $$: [tmp0, tmp1, ...]}`.

The `.pointers` property is now a proxy to allow the rest of the VM and primitives to still work. The JIT and optimized primitives will not use this proxy.

codefrau committed:
Inst var fix

codefrau committed:
Fix 64bit inst vars

codefrau commented:
I implemented actual var names, see b6a32befdc537ef28431378b9c1971480be3c086

The VM still works thanks to the `.pointers` now being a proxy to those explicit int vars. 

I tried it with various images (old, new, 32-bit, 64-bit) but it may be helpful if you could try others too. Everything should work as before, just slower.

gkgoat1 commented:
Could we port v2 to ESM and JS-native classes, considering as, iirc, Lively4 and Lively Next support it, in addition to most non-Lively coding environments and tools?

codefrau commented:
> Could we port v2 to ESM and JS-native classes, considering as, iirc, Lively4 and Lively Next support it, in addition to most non-Lively coding environments and tools?

I'm not sure how much we would gain from that. We would lose the ability to use the Lively tools at Dan's [Smalltalk Zoo](https://smalltalkzoo.thechm.org/squeak.html) though.

So I'm not entirely opposed but it's definitely not a priority for me since the behavior wouldn't change at all and v2 is about better behavior.

gkgoat1 commented:
> > Could we port v2 to ESM and JS-native classes, considering as, iirc, Lively4 and Lively Next support it, in addition to most non-Lively coding environments and tools?
> 
> I'm not sure how much we would gain from that. We would lose the ability to use the Lively tools at Dan's [Smalltalk Zoo](https://smalltalkzoo.thechm.org/squeak.html) though.
That might present an issue, unless Smalltalk Zoo is updated.
> 
> So I'm not entirely opposed but it's definitely not a priority for me since the behavior wouldn't change at all and v2 is about better behavior.
My particular issue is about the usage of globals and extension methods, which make SqueakJS hard to optimize, especially AOT, and hard to scope (what if there are *two* SqueakJSs on one page?).

codefrau commented:
> > We would lose the ability to use the Lively tools at Dan's [Smalltalk Zoo](https://smalltalkzoo.thechm.org/squeak.html) though.
> 
> That might present an issue, unless Smalltalk Zoo is updated.

Dan is not getting any younger, I doubt that would happen. 

> > So I'm not entirely opposed but it's definitely not a priority for me since the behavior wouldn't change at all and v2 is about better behavior.
> 
> My particular issue is about the usage of globals and extension methods, which make SqueakJS hard to optimize, especially AOT, and hard to scope (what if there are *two* SqueakJSs on one page?). 

The majority of run time will be spent inside of generated code once the new JIT is in place. I don't think spending much effort to optimize the support code is worth the trouble. 

BTW what AOT are you envisioning?

gkgoat1 commented:
> > > We would lose the ability to use the Lively tools at Dan's [Smalltalk Zoo](https://smalltalkzoo.thechm.org/squeak.html) though.
> > 
> > 
> > That might present an issue, unless Smalltalk Zoo is updated.
> 
> Dan is not getting any younger, I doubt that would happen.
> 
> > My particular issue is about the usage of globals and extension methods, which make SqueakJS hard to optimize, especially AOT, and hard to scope (what if there are _two_ SqueakJSs on one page?).
> 
> The majority of run time will be spent inside of generated code once the new JIT is in place. I don't think spending much effort to optimize the support code is worth the trouble.
> 
> BTW what AOT are you envisioning?

Compiling SqueakJS, with a "factory" image, into an appropriate binary with [ComponentizeJS](https://github.com/bytecodealliance/ComponentizeJS) or similar tools (for similar purposes as Sugar's EToys integration for my projects)

codefrau commented:
> Compiling SqueakJS, with a "factory" image, into an appropriate binary with [ComponentizeJS](https://github.com/bytecodealliance/ComponentizeJS) or similar tools (for similar purposes as Sugar's EToys integration for my projects)

I'm sorry, targeting WebAssembly in that way is madness and would result in a much slower execution and a more bloated download. I could elaborate but not on here. We need to take this discussion elsewhere because it's getting very much away from SqueakJS V2. Please continue on the mailing list, preferably vm-dev. If you don't like that, open a separate ticket (but you would get more people to engage on a mailing list)

codefrau committed:
lint

codefrau committed:
Fix pointsTo

codefrau committed:
Merge branch 'main' into v2

codefrau committed:
Post-merge fixups

codefrau committed:
Don't use actual inst var names but numeric suffix for each

codefrau committed:
Use direct inst vars in jit

codefrau committed:
Fix dirty marking in jit

codefrau committed:
Update README.md

codefrau commented:
> are instance variable names a concept that the VM should be aware of at all?

After thinking this over more you are right, the VM should not be aware of inst var names. I now in this branch implemented a scheme like `rcvr.$0`, `rcvr.$1`, etc. and the variable-size pointers are in an array called `rcvr.$$`.

However, working with this is rather impractical, and no fun (named vars would at least have been fun). 

I tried converting just the interpreter's Context access to use these properties, but ran into issues where throughout the VM, it's assumed that inst vars and indexed vars are contiguous in the `pointers` array. That means the index differs between `$$` and `pointers` which would have to be fixed everywhere before anything works again. I got quite far but couldn't figure out where it went wrong.

In the end I decided to abandon this approach. The performance gain of a few percent with explicit inst vars is not worth the hassle, and there would be performance losses too whenever the inst vars need to be iterated over (like in object allocation, GC, snapshotting, etc). I'm now convinced that I should keep the object model as is.
